discv5/discv5-wire-test-vectors.md

```diff
@@ -0,0 +1,338 @@
+
+The three keys, resulting from the HKDF-EXPAND function.
+
+    initiator-key: 0x238d8b50e4363cf603a48c6cc3542967
+    recipient-key: 0xbebc0183484f7e7ca2ac32e3d72c8891
+    auth-resp-key: 0xe987ad9e414d5b4f9bfe4ff1e52f2fae
+
+### Nonce Signing
+
+Nonce signatures should prefix the string `discovery-id-nonce` and post-fix the
+ephemeral key before taking the `sha256` hash of the `id-nonce`.
+
+#### Input Parameters
+
+The `local-secret-key` is the raw secp256k1 private key used to sign the nonce.
+
+    id_nonce: 0x02a77e3aa0c144ae7c0a3af73692b7d6e5b7a2fdc0eda16e8d5e6cb0d08e88dd04
+    ephemeral-key: 0x9961e4c2356d61bedb83052c115d311acb3a96f5777296dcf297351130266231503061ac4aaee666073d7e5bc2c80c3f5c5b500c1cb5fd0a76abbb6b675ad157
+    local-secret-key: 0xfb757dc581730490a1d7a00deea65e9b1936924caaea8f44d476014856b68736
+
+#### Expected Outputs
+
+    id-nonce-sig: 0x3b7b8ce9df3fbd9b6367c365622ccc82a2cb9d94219401e7b08e3194f9f835764a07caad38bf0f5a7a89501a8156bb053c880774502f5cd8a6190fbe374adc89
+
+### Encryption/Decryption
+
+This test vector demonstrates the `AES_GCM` encryption/decryption used in the wire
+protocol.
+
+#### Input Parameters
+
+    encryption-key: 0x9f2d77db7004bf8a1a85107ac686990b
+    nonce: 0x27b5af763c446acd2749fe8e
+    pt: 0x01c20101
+    ad: 0x93a7400fa0d6a694ebc24d5cf570f65d04215b6ac00757875e3f3a5f42107903
+
+#### Expected Output
+
+Note that the 16 byte MAC is prepended to the ciphertext.
+
+    message-ciphertext: a5d12a2d94b8ccb3ba55558229867dc13bfa3648
+
+### Authentication Header and Encrypted Message Generation
+
+This test demonstrates the construction of an authentication header. Given a
+local private key, a message `tag`, the local `node-id`, an `id-nonce`, an
+encryption key and a protocol message to send, one should be able to build an
+encrypted message with an authentication header.
+
+This test is broken down into stages.
+
+1. The generation of the rlp-encoded `auth-pt` which is required in the
+   authentication header
+2. The Encryption of `auth-pt`
+3. The generation of the authentication header itself
+4. Encrypting the message and combining into an RLP-encoded `message-packet`
+   with an `auth-header`
+
+#### Part 1: Auth-pt Generation
+
+This first section entails signature generation, and adding any ENR into
+`auth-pt`. In this example, there is no ENR sent. This tests the signature
+generation and correct RLP encoding of the `auth-pt` before encryption.
+
+##### Input Parameters
+
+    secret-key: 0x7e8107fe766b6d357205280acf65c24275129ca9e44c0fd00144ca50024a1ce7
+    id-nonce: 0xe551b1c44264ab92bc0b3c9b26293e1ba4fed9128f3c3645301e8e119f179c65
+    ephemeral-pubkey: 0xb35608c01ee67edff2cffa424b219940a81cf2fb9b66068b1cf96862a17d353e22524fbdcdebc609f85cbd58ebe7a872b01e24a3829b97dd5875e8ffbc4eea81
+    enr: []
+
+##### Expected Output
+
+    auth-pt: 0xf84405b840f753ac31b017536bacd0d0238a1f849e741aef03b7ad5db1d4e64d7aa80689931f21e590edcf80ee32bb2f30707fec88fb62ea8fbcd65b9272e9a0175fea976bc0
```

Comment:
As idNonceSig is incorrect, this one is too.
`0xf84405b840f753ac31b017536bacd0d0238a1f849e741aef03b7ad5db1d4e64d7aa80689931f21e590edcf80ee32bb2f30707fec88fb62ea8fbcd65b9272e9a0175fea976b80` for me

Reply:
Signature value depends on how signing is implemented. The output will match only if your library implements deterministic signatures, but that's not a requirement.

Reply:
This is not an error due to nonce-sig mismatch. In fact, we agree on the nonce sig. The difference between our results is the rlp encoding of an empty ENR. You result is encoding an empty enr as an empty string `0x80`, whereas I (and I assume geth) are encoding it as an empty list [] `0xc0`. I assume the empty list is the correct encoding.

Reply:
@AgeManning nice catch, fixed. Everything matches for me since now.